text/3-entrees/babe-grandpa-node.md

```diff
@@ -0,0 +1,187 @@
+	force_authoring,
+	babe_link,
+	can_author_with,
+};
+```
+
+With the parameters established, we can now create and spawn the authorship future.
+
+```rust, ignore
+let babe = sc_consensus_babe::start_babe(babe_config)?;
+service.spawn_essential_task("babe", babe);
+```
+
+## Spawning the Grandpa Task
+
+Just as we needed an `async` worker to author blocks with Babe, we need an `async` worker to listen
+to and cast grandpa votes. Again, we begin by creating an instance of
+[`GrandpaParams`](https://crates.parity.io/sc_finality_grandpa/struct.GrandpaParams.html)
+
+```rust, ignore
+let grandpa_config = sc_finality_grandpa::GrandpaParams {
+	config: grandpa_config,
+	link: grandpa_link,
+	network: service.network(),
+	inherent_data_providers: inherent_data_providers.clone(),
+	telemetry_on_connect: Some(service.telemetry_on_connect_stream()),
+	voting_rule: sc_finality_grandpa::VotingRulesBuilder::default().build(),
+	prometheus_registry: service.prometheus_registry(),
+};
+```
+
+With the parameters established, we can now create and spawn the authorship future.
+
+```rust, ignore
+service.spawn_essential_task(
+	"grandpa-voter",
+	sc_finality_grandpa::run_grandpa_voter(grandpa_config)?
+);
+```
+
+### Disabled Grandpa
+
+Proof of Authority networks generally contain many full nodes that are not authorities. When Grandpa
+is present in the network, we still need to tell the node how to interpret grandpa-related messages
+it may receive (just ignore them) and ensure that the correct inherents are still included in blocks
+in the case that the node _is_ an authority in Babe but not Grandpa.
+
+```rust, ignore
+sc_finality_grandpa::setup_disabled_grandpa(
+	service.client(),
+	&inherent_data_providers,
+	service.network(),
+)?;
+```
+
+## Constraints on the Runtime
+
+### Runtime APIs
+
+Both Babe and Grandpa rely on getting their authority sets from the runtime via the
+[BabeAPI](https://crates.parity.io/sc_consensus_babe/trait.BabeApi.html) and the
+[GrandpaAPI](https://crates.parity.io/sp_finality_grandpa/trait.GrandpaApi.html). So trying to build
+this node with a runtime that does not provide these APIs will fail to compile.
+
+### Pre Runtime Digests
+
+Just as we cannot use this node with a runtime that does not provide the appropriate runtime APIs,
+we also cannot use a runtime designed for this node with different consensus engines.
+
+Because BABE is a slot-based consensus engine, it must inform the runtime which slot each block was
+intended for. To do this, it uses a technique known as a pre-runtime digest. It has two kinds,
+[`PrimaryPreDigest`](https://crates.parity.io/sc_consensus_babe/struct.PrimaryPreDigest.html) and
+[`SecondaryPlainPreDigest`](https://crates.parity.io/sc_consensus_babe/struct.SecondaryPlainPreDigest.html).
+The Babe authorship task automatically inserts these digest items in each block it authors.
+
+Because the runtime needs to interpret these pre-runtime digests, they are not optional. That means
```

Comment:
I am not sure about this, in practice the only reusability we lose is that in order to use the BABE node code you need to have the BABE pallet in your runtime (likewise for GRANDPA). I'd say that is expected and OK. The rest of the modules of the runtime are unaffected and don't need to know about any particularity of consensus or pre-digests.

Reply:
I agree with everything you said here. I'm trying to explain to readers why the `babe-grandpa-runtime` isn't an option available to use in the pow node or the instant seal node like all the other runtimes in the recipes are. I agree it is a straight-forward change to the runtime to remove the BABE pallet, but it is necessary change. Maybe this rephrasing makes that more clear.

> That means runtimes that expect the pre-digests _require_ the pre-digests, and cannot be used unmodified in nodes that don't provide the pre-digests. Unlike other runtimes in the Recipes where
runtimes can be freely swapped between nodes, the `babe-grandpa-runtime` can _only_ be used in a
node that is actually running BABE

Reply:
It's also worth mentioning that really the only interesting thing about the Recipes' `babe-grandpa-runtime` is that it has these consensus pallets.